### PR TITLE
Manifold gsd

### PR DESCRIPTION
**Pull request by mthran**
## Description

This PR implements `append_microstate` for `Spherical<3>`, `Spherical<4>`, `Hyperbolic<3>`, and `OrientedHyperbolicPoint<3, Angle>`. 

## Motivation and context

`append_microstate` is necessary to save Monte Carlo trajectories as GSD files. The four aforementioned manifold types are currently the four supported non-Euclidean manifold types. 

## How has this been tested?

This PR includes unit tests for each of the four manifold types with every relevant combination of boundary conditions. I have also been using these functions for production simulations and verified that they work as intended. 

## Checklist:

- [ X ] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-rs/blob/trunk/CONTRIBUTING.md).
- [ X ] I agree with the terms of the [**hoomd-rs Contributor Agreement**](https://github.com/glotzerlab/hoomd-rs/blob/trunk/ContributorAgreement.md).
- [ X ] My name is on the list of contributors (`doc/src/credits.md`) in the pull request source branch.
- [x] I have summarized these changes in `release-notes.md` following the established format.

### CODE REVIEW

File: hoomd-microstate/src/append.rs [hunk 1]
```diff
@@ -202,15 +203,97 @@ impl<B, X> AppendMicrostate<B, OrientedPoint<Cartesian<3>, Versor>, X, Periodic<
...
+    fn append_microstate(
+        &mut self,
+        microstate: &Microstate<B, Point<Spherical<3>>, X, C>,
+    ) -> Result<Frame<'_>, AppendError> {
+        self.append_frame(microstate.step())?
+            .configuration_box([5.0, 5.0, 5.0, 0.0, 0.0, 0.0])?
+            .configuration_dimensions(Dimensions::Three)?
+            .particles_position(microstate.iter_sites_tag_order().map(|s| {
+                let pt = *s.properties.position.coordinates();
+                pt.into()
```
Comment (joaander) [inline hoomd-microstate/src/append.rs]:
```suggestion
                let pt = *s.properties.position.point()
```
This has fewer levels of indirection.

File: hoomd-microstate/src/append.rs [hunk 2]
```diff
@@ -202,15 +203,97 @@ impl<B, X> AppendMicrostate<B, OrientedPoint<Cartesian<3>, Versor>, X, Periodic<
...
 }
 
+impl<B, X, C> AppendMicrostate<B, Point<Spherical<3>>, X, C> for HoomdGsdFile {
+    #[inline]
+    fn append_microstate(
+        &mut self,
+        microstate: &Microstate<B, Point<Spherical<3>>, X, C>,
+    ) -> Result<Frame<'_>, AppendError> {
+        self.append_frame(microstate.step())?
+            .configuration_box([5.0, 5.0, 5.0, 0.0, 0.0, 0.0])?
```
Comment (joaander) [inline hoomd-microstate/src/append.rs]: Why 5,5,5? Shouldn't it be 1,1,1 to cover the unit sphere?

File: hoomd-microstate/src/append.rs [hunk 3]
```diff
@@ -202,15 +203,97 @@ impl<B, X> AppendMicrostate<B, OrientedPoint<Cartesian<3>, Versor>, X, Periodic<
...
+}
+
+impl<B, X, C> AppendMicrostate<B, Point<Spherical<4>>, X, C> for HoomdGsdFile {
+    #[inline]
+    fn append_microstate(
+        &mut self,
+        microstate: &Microstate<B, Point<Spherical<4>>, X, C>,
+    ) -> Result<Frame<'_>, AppendError> {
+        self.append_frame(microstate.step())?
+            .configuration_box([5.0, 5.0, 5.0, 0.0, 0.0, 0.0])?
```
Comment (joaander) [inline hoomd-microstate/src/append.rs]: Why 5,5,5?

File: hoomd-microstate/src/append.rs [hunk 4]
```diff
@@ -202,15 +203,97 @@ impl<B, X> AppendMicrostate<B, OrientedPoint<Cartesian<3>, Versor>, X, Periodic<
...
+
+impl<B, X, C> AppendMicrostate<B, OrientedHyperbolicPoint<3, Angle>, X, C> for HoomdGsdFile {
+    #[inline]
+    fn append_microstate(
+        &mut self,
+        microstate: &Microstate<B, OrientedHyperbolicPoint<3, Angle>, X, C>,
+    ) -> Result<Frame<'_>, AppendError> {
+        self.append_frame(microstate.step())?
+            .configuration_box([2.0, 2.0, 2.0, 0.0, 0.0, 0.0])?
+            .configuration_dimensions(Dimensions::Two)?
```
Comment (joaander) [inline hoomd-microstate/src/append.rs]:
```suggestion
            .configuration_box([2.0, 2.0, 0.0, 0.0, 0.0, 0.0])?
            .configuration_dimensions(Dimensions::Two)?
```
Inconsistent as written. Should either be 2,2,0 and `Dimensions::Two` or 2,2,2 and `Dimensions::Three`.

File: hoomd-microstate/src/append.rs [hunk 5]
```diff
@@ -202,15 +203,97 @@ impl<B, X> AppendMicrostate<B, OrientedPoint<Cartesian<3>, Versor>, X, Periodic<
...
+
+impl<B, X, C> AppendMicrostate<B, Point<Hyperbolic<3>>, X, C> for HoomdGsdFile {
+    #[inline]
+    fn append_microstate(
+        &mut self,
+        microstate: &Microstate<B, Point<Hyperbolic<3>>, X, C>,
+    ) -> Result<Frame<'_>, AppendError> {
+        self.append_frame(microstate.step())?
+            .configuration_box([2.0, 2.0, 2.0, 0.0, 0.0, 0.0])?
+            .configuration_dimensions(Dimensions::Two)?
```
Comment (joaander) [inline hoomd-microstate/src/append.rs]:
```suggestion
            .configuration_box([2.0, 2.0, 0.0, 0.0, 0.0, 0.0])?
            .configuration_dimensions(Dimensions::Two)?
```

**Review by joaander — COMMENTED** [review]
Thanks! This looks great, it is good to see complete unit tests. I have some questions and offer suggetions in the line-specific comments below.

Please also add a new section to the `AppendMicrostate` trait documentation (https://hoomd-rs.readthedocs.io/en/1.1.0/api/hoomd_microstate/trait.AppendMicrostate.html) that lists which projection is used for each curved space.

**Review by joaander — APPROVED** [review]
Thanks!
